std/machines/hash/keccakf16.asm

```diff
@@ -0,0 +1,576 @@
+    //     pub export: T,
+
+    //     /// Permutation inputs, stored in y-major order.
+    //     pub preimage: [[[T; U64_LIMBS]; 5]; 5],
+
+    //     pub a: [[[T; U64_LIMBS]; 5]; 5],
+
+    //     /// ```ignore
+    //     /// C[x] = xor(A[x, 0], A[x, 1], A[x, 2], A[x, 3], A[x, 4])
+    //     /// ```
+    //     pub c: [[T; 64]; 5],
+
+    //     /// ```ignore
+    //     /// C'[x, z] = xor(C[x, z], C[x - 1, z], C[x + 1, z - 1])
+    //     /// ```
+    //     pub c_prime: [[T; 64]; 5],
+
+    //     // Note: D is inlined, not stored in the witness.
+    //     /// ```ignore
+    //     /// A'[x, y] = xor(A[x, y], D[x])
+    //     ///          = xor(A[x, y], C[x - 1], ROT(C[x + 1], 1))
+    //     /// ```
+    //     pub a_prime: [[[T; 64]; 5]; 5],
+
+    //     /// ```ignore
+    //     /// A''[x, y] = xor(B[x, y], andn(B[x + 1, y], B[x + 2, y])).
+    //     /// ```
+    //     pub a_prime_prime: [[[T; U64_LIMBS]; 5]; 5],
+
+    //     /// The bits of `A''[0, 0]`.
+    //     pub a_prime_prime_0_0_bits: [T; 64],
+
+    //     /// ```ignore
+    //     /// A'''[0, 0, z] = A''[0, 0, z] ^ RC[k, z]
+    //     /// ```
+    //     pub a_prime_prime_prime_0_0_limbs: [T; U64_LIMBS],
+    // }
+
+    pol commit preimage[5 * 5 * 4];
+    pol commit a[5 * 5 * 4];
+    pol commit c[5 * 64];
+    pol commit c_prime[5 * 64];
+    pol commit a_prime[5 * 5 * 64];
+    pol commit a_prime_prime[5 * 5 * 4];
+    pol commit a_prime_prime_0_0_bits[64];
+    pol commit a_prime_prime_prime_0_0_limbs[4];
+
+    // Initially, the first step flag should be 1 while the others should be 0.
+    // builder.when_first_row().assert_one(local.step_flags[0]);
+    // for i in 1..NUM_ROUNDS {
+    //     builder.when_first_row().assert_zero(local.step_flags[i]);
+    // }
+    // for i in 0..NUM_ROUNDS {
+    //     let current_round_flag = local.step_flags[i];
+    //     let next_round_flag = next.step_flags[(i + 1) % NUM_ROUNDS];
+    //     builder
+    //         .when_transition()
+    //         .assert_eq(next_round_flag, current_round_flag);
+    // }
+
+    let step_flags: col[NUM_ROUNDS] = array::new(NUM_ROUNDS, |i| |row| if row % NUM_ROUNDS == i { 1 } else { 0 } );
+
+    // let main = builder.main();
+    // let (local, next) = (main.row_slice(0), main.row_slice(1));
+    // let local: &KeccakCols<AB::Var> = (*local).borrow();
+    // let next: &KeccakCols<AB::Var> = (*next).borrow();
+
+    // let first_step = local.step_flags[0];
+    // let final_step = local.step_flags[NUM_ROUNDS - 1];
+    // let not_final_step = AB::Expr::one() - final_step;
+
+    let first_step: expr = step_flags[0]; // aliasing instead of defining a new fixed column
+    let final_step: expr = step_flags[NUM_ROUNDS - 1];
+    // Equals to 1 for all non-last rows in all 24-row blocks. Forced to zero for the very last row of the submachine (if partial block).
+    col fixed active_row = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0]* + [0];
```

Comment:
I'm not sure about this name. How about `internal_row`? Or `in_progress`?

If I understand this right, the only purpose is to activate a copy from some result in the current row to the input of the next row. Which we don't want to happen in the last row of the block *and* the last row of the trace. Think this could be explained in the comment.

Reply:
Yes. That's the right idea behind this column.

Reply:
Done